### PR TITLE
[CDF-22608] Add data set support for data workflows

### PR DESCRIPTION
**Pull request by VerstraeteBert**
## Description
Add data set support for data workflows on both the groups API and Data Workflows API. The capability lives on the Workflow object; any operations on versions, executions, triggers, tasks belonging to or spawned by that workflow require data set access to the data set of the workflow.

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/data_classes/workflows.py [hunk 1]
```diff
@@ -49,8 +50,11 @@ class WorkflowUpsert(WorkflowCore):
     Args:
         external_id (str): The external ID provided by the client. Must be unique for the resource type.
         description (str | None): Description of the workflow. Note that when updating a workflow, the description will
-                            always be overwritten also if it is set to None. Meaning if the wokflow already has a description,
+                            always be overwritten also if it is set to None. Meaning if the workflow already has a description,
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:53]: _wok flow_ is like, so much cooler

File: cognite/client/data_classes/workflows.py [hunk 2]
```diff
@@ -49,8 +50,11 @@ class WorkflowUpsert(WorkflowCore):
...
                             and you want to keep it, you need to provide the description when updating the workflow.
+        data_set_id (int | None): The id of the data set this workflow belongs to.
+                            If a dataSetId is provided, any operations on this workflow, or its versions, executions,
+                            and triggers will require appropriate access to the data set.
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:57]: Adding dataset seems quite complex to me. Let's say it is write-protected, will that stop someone with only read access from triggering for example? It seems worth it to change **...will require appropriate access...** to a more verbose explanation! 😄

File: cognite/client/data_classes/workflows.py [hunk 3]
```diff
@@ -88,13 +93,15 @@ def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> S
             external_id=resource["externalId"],
             description=resource.get("description"),
             created_time=resource["createdTime"],
+            data_set_id=resource["dataSetId"],
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:96]:
```suggestion
            data_set_id=resource.get("dataSetId"),
```

File: tests/tests_integration/test_api/test_data_workflows.py [hunk 4]
```diff
@@ -30,9 +30,11 @@
 
 @pytest.fixture
 def workflow_list(cognite_client: CogniteClient) -> WorkflowList:
+    data_set = cognite_client.data_sets.list(limit=1)[0]
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_data_workflows.py:33]: Good candidate for a fixture

File: cognite/client/data_classes/capabilities.py [hunk 5]
```diff
@@ -1236,6 +1236,7 @@ class Action(Capability.Action):  # type: ignore [misc]
 
     class Scope:
         All = AllScope
+        DataSet = DataSetScope
```
Comment (haakonvt) [inline cognite/client/data_classes/capabilities.py:1239]: Please add a test in
`tests/tests_unit/test_data_classes/test_capabilities.py`

File: cognite/client/data_classes/workflows.py (hunk 1 above)
Comment (VerstraeteBert) [reply cognite/client/data_classes/workflows.py:53]: 🍜

File: cognite/client/data_classes/workflows.py (hunk 2 above)
Comment (VerstraeteBert) [reply cognite/client/data_classes/workflows.py:57]: I'm not sure if this belongs here. Data set behavior should be the same for different resources and has its own documentation. I've documented where it's unique: workflows are assigned to data sets, sub-resources or linked ones require the same access.

Comment (VerstraeteBert) [reply cognite/client/data_classes/workflows.py]: I've added a link to the datasets docs, which does explain the more subtle concepts.

File: tests/tests_integration/test_api/test_data_workflows.py [hunk 6]
```diff
@@ -266,9 +265,13 @@ def workflow_scheduled_trigger(cognite_client: CogniteClient, add_multiply_workf
     cognite_client.workflows.triggers.delete(trigger.external_id)
 
 
+@pytest.fixture(scope="session")
+def data_set(cognite_client: CogniteClient) -> DataSet:
+    return cognite_client.data_sets.list(limit=1)[0]
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_data_workflows.py:270]: I guess this is safe (`IndexError`) as data sets can't be deleted 😂

File: CHANGELOG.md [hunk 7]
```diff
@@ -17,6 +17,11 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [7.64.12] - 2023-11-12
+### Added
+- Added Scope `DataSet` for `WorkflowOrchestrationAcl`.
+- Added optional `data_set_id` to `Workflow` and `WorkflowUpsert`.
```
Comment (haakonvt) [inline CHANGELOG.md:23]: Just a suggestion, I do like how precise you are with the added features, but the overall important message is imo that workflows now support data sets:
```suggestion
## [7.64.12] - 2023-11-14
### Added
- Workflows now support data sets
```

**Review by haakonvt — APPROVED** [review]
LGTM!

File: CHANGELOG.md (hunk 7 above)
Comment (VerstraeteBert) [reply CHANGELOG.md:23]: Fair enough!

**Pull request merged**